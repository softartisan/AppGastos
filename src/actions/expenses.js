import uuid from 'uuid';
import database from '../firebase/firebase';

//ADD_EXPENSE
const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
    description = '',
    note ='',
    amount = 0,
    createdAt = 0 
    } = expenseData;

    const expense = {
      description,
      note,
      amount,
      createdAt
    }

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id:ref.key,
        ...expense
      }));
    });

  }
}



//REMOVE_EXPENSE
const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    idToRemove: id
});

const startRemoveExpense = ({id} = {}) => {

  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}));
    })
  }

}


//EDIT_EXPENSE
const editExpense = (id,updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

const startEditExpense = (id, updates) => {
// database.ref().update({
//   stressLevel: 9,
//   'job/company': 'Amazon',
//   'location/city': 'Seattle'
// });
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update({
      ...updates
    }).then(() => {
      dispatch(editExpense(id, updates));
    })
  }

}

//SET_EXPENSES
const setExpenses = (expenses) => {
  return { type: 'SET_EXPENSES', expenses }
}

const startSetExpenses = () => {
  //Retorno una funciÃ³n
  return (dispatch) => {
    //Saco expenses desde firebase y las guardo en expenses
    //Retorno esta Promise
    return database.ref('expenses').once('value').then((snapshot) => {

      const expenses = [];
      //Por cada objeto de firebase lleno un objeto del array expenses
      snapshot.forEach((childSnapshot) => {
        expenses.push( {id: childSnapshot.key, ...childSnapshot.val()} );
      });

      dispatch(setExpenses(expenses));

    });
  }
}

export {
    startEditExpense,
    editExpense,
    removeExpense,
    startRemoveExpense,
    addExpense,
    startAddExpense,
    setExpenses,
    startSetExpenses
}