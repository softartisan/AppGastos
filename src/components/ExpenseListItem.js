import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({dispatch, id, description, amount, createdAt}) => (
    <div>
        <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
        </Link>
        <p>
        {numeral(amount).format('$000,000,000')} 
        - 
        {moment(createdAt).format('DD MMMM YYYY')}
        </p>
    </div>
);

export default ExpenseListItem;