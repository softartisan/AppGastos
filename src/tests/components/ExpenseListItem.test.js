import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('Should render ListItem', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
  expect(toJSON(wrapper)).toMatchSnapshot();
});