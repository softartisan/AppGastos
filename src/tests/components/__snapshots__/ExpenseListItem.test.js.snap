// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ListItem 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      cafe
    </h3>
  </Link>
  <p>
    $100
    -
    31 December 1969
  </p>
</div>
`;
